### ISSUE TITLE
DerivedArtifact should expose a getExtension()

### ISSUE THREAD

**Issue by loosebazooka** [gradle/gradle#28969]
### Expected Behavior

A signer alternative to pgp (like sigstore in this case) which uses an extension `.sigstore.json` should be able to publish with that extension when doing a maven publish (this is not supported without a bit of a hack).

The [PublicationInternal](https://github.com/gradle/gradle/blob/ec838d879d29c73f4cefe628deed2902bd67bc06/platforms/software/publish/src/main/java/org/gradle/api/publish/internal/PublicationInternal.java#L81) object could add the extension to the interface
```diff
    interface DerivedArtifact {
        boolean shouldBePublished();
        File create();
+       String getExtension();
    }
```

I also understand that we kinda need to access some of these internal APIs to make this happen from a 3rd party plugin.

### Current Behavior (optional)

The extension is computed from the filename which means it's the last of `X.Y.Z` which in this case is `.Z` instead of `.Y.Z`. See code: https://github.com/gradle/gradle/blob/ec838d879d29c73f4cefe628deed2902bd67bc06/platforms/software/maven/src/main/java/org/gradle/api/publish/maven/internal/artifact/DerivedMavenArtifact.java#L44

So during publication of `my.jar.sigstore.json`, the maven publication generates a finished dervied object as `my.jar.json` (instead of `my.jar.sigstore.json`)

### Context

Sigstore signatures are expected to use the final extension `.sigstore.json`. There is a workaround that can be done by explicitly setting the extension on the PublicationArtifact (which we do infact exploit), but the current interface for DerviedArtifact could make this less hacky.

**Comment by ov7a** [comment]
This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

***

Relates to:
* https://github.com/gradle/gradle/issues/26760

**Comment by big-guy** [comment]
@loosebazooka do you have a workaround for now or would this unblock you?

**Comment by loosebazooka** [comment]
Yeah workaround exists. Plus I would have to be pretty backward compatible for now. This is just a request for a change in the future.

**Comment by big-guy** [comment]
We should revisit this for 9.x so that we support sigstore more properly.
